ldap/servers/plugins/acl/acl.c

```diff
@@ -2408,10 +2466,53 @@ acl__resource_match_aci(Acl_PBlock *aclpb, aci_t *aci, int skip_attrEval, int *a
                                                     ACL_EVAL_TARGET_FILTER);
             slapi_ch_free((void **)&lasinfo);
         } else {
-            if (slapi_vattr_filter_test(NULL, aclpb->aclpb_curr_entry,
-                                        aci->targetFilter,
-                                        0 /*don't do access check*/) != 0) {
-                filter_matched = ACL_FALSE;
+            Slapi_DN *sdn;
+            char* attr_evaluated = "None";
+            char logbuf[1024];
+            char *filterstr; /* key to retrieve/add targetfilter value in the cache */
+            struct targetfilter_cached_result *previous_filter_test;
+            
+            /* only usefull for debug purpose */
```

Comment:
If its for debug could this be in a #DEBUG define?

Reply:
@Firstyear many thanks also for having look at this PR.
ATM this PR is kind of preliminary and I would also appreciate any feedback on the use of a cache mechanism itself. ACI is a key security component and I am a bit concerned about caching anything there.
I think the cache is safe as it is at a per entry level. So when we evaluate a filter against an entry , the result of this evaluation is still valid for the same filter against the same entry again.

A concern is how do we know it is the same filter ? is strcmp valid for that ?

Regarding the hit rate, it depends on the ACI target definition and how much attributes are returned. In my tests, the SRCH returned 8 attributes and the ACI, containing the targetfilter, target all those attributes. So instead of evaluating 9 times the targetfilter (8 attributes + 1entry) it evaluates it only once. If the targetfilter is complex like 
`(&(objectclass=hosts)(fqdn=*.domain.us)(name=*foo)(description=*bla bla bla)....(title=*xxx))` and expensive to compute like substring matching rules, then extra evaluations have a significant impact. On a testcase retrieving 500 entries with 8 attributes and one targetfilter ACI it goes from 4500 evaluations without the cache to 500 with the cache. If there are several selected ACIs with targetfilters it makes things worse.
From a test etime dropped from 2.5s to 0.5s.

Reply:
"A concern is how do we know it is the same filter ? is strcmp valid for that ?" -- Yep, that's find in this case. 

And that's a good improvement there, I didn't know how it was evaluated multiple times like that :)